host.go

```diff
@@ -289,7 +289,32 @@ func (h *gpbftRunner) receiveCertificate(c *certs.FinalityCertificate) error {
 	log.Infow("skipping forwards based on cert", "from", currentInstance, "to", nextInstance)
 
 	nextInstanceStart := h.computeNextInstanceStart(c)
-	return h.participant.StartInstanceAt(nextInstance, nextInstanceStart)
+	return h.startInstanceAt(nextInstance, nextInstanceStart)
+}
+
+func (h *gpbftRunner) startInstanceAt(instance uint64, at time.Time) error {
+	// Look for any existing messages in WAL for the next instance, and if there is
+	// any replay them to self to aid the participant resume progress when possible.
+	//
+	// Collect the replay messages first then replay them to avid holding the lock
+	// for too long if for whatever reason validate/receive is slow.
+	var replay []*gpbft.GMessage
+	h.msgsMutex.Lock()
+	if messages, found := h.selfMessages[instance]; found {
+		replay = make([]*gpbft.GMessage, 0, len(messages))
+		for _, message := range messages {
+			replay = append(replay, message...)
+		}
+	}
+	h.msgsMutex.Unlock()
+	for _, message := range replay {
+		if validated, err := h.participant.ValidateMessage(message); err != nil {
+			log.Warnw("invalid self message", "message", message, "err", err)
+		} else if err := h.participant.ReceiveMessage(validated); err != nil {
+			log.Warnw("failed to send resumption message", "message", message, "err", err)
```

Comment:
The only thing I'm worried is the ordering of messages but it is probably fine.

Reply:
It should just work, but no harm in sorting them for a more optimal resumption 👍

Reply:
OK added